File: shared/BuddyCore/Sources/BuddyCore/Domain/RacerPhysics.swift
import Foundation

/// Pure deterministic 2D top-down racer physics. `tick(state, dt, input)`
/// returns the next state. Identical inputs + identical dt produce identical
/// state on both peers, enabling client-side prediction + host reconciliation.
public enum RacerPhysics: GameStateReducer {
    public typealias State = RacerState
    public typealias Input = RacerInput

    /// Track dimensions. v1 ships a single rectangular arena.
    public static let trackWidth: Double  = 100.0
    public static let trackHeight: Double = 60.0

    /// Physics constants — tuned for "fun > realistic".
    public static let maxSpeed: Double      = 30.0
    public static let acceleration: Double  = 50.0
    public static let braking: Double       = 70.0
    public static let idleDecay: Double     = 12.0
    public static let turnRate: Double      = 3.5  // rad/s
    public static let bounceFactor: Double  = 0.55

    public static func initialState(host: Peer, guest: Peer) -> RacerState {
        RacerState(
            host: host.id,
            guest: guest.id,
            cars: [
                host.id:  RacerCar(x: 20, y: 30, heading: 0, speed: 0),
                guest.id: RacerCar(x: 80, y: 30, heading: .pi, speed: 0),
            ],
            laps: [host.id: 0, guest.id: 0],
            tickCount: 0,
            winnerId: nil
        )
    }

    /// `reduce` adapts the protocol API: apply the input, advance one fixed
    /// 33 ms tick, and surface the winner if one was just declared.
    public static func reduce(_ state: RacerState, input: RacerInput) throws -> Step {
        if state.winnerId != nil { throw Error.gameOver }
        var s = state
        s.cars[input.player]?.lastInput = input
        let next = tick(s, dtMillis: 33)
        let outcome: Outcome? = next.winnerId.map { .winner($0) }
        return (next, outcome)
    }

    public static func isFinal(_ state: RacerState) -> Bool {
        state.winnerId != nil
    }

    /// Real-time game — no `currentTurn`.
    public static func currentTurn(_ state: RacerState) -> UUID? { nil }

    // MARK: - Tick

    /// Fixed-timestep physics step. Used directly by the host's game loop;
    /// the guest replays it for client-side prediction.
    public static func tick(_ state: RacerState, dtMillis: Int) -> RacerState {
        if state.winnerId != nil { return state }
        let dt = Double(dtMillis) / 1000.0
        var s = state
        for (id, car) in s.cars {
            s.cars[id] = step(car, dt: dt)
        }
        s.tickCount += 1
        return s
    }

    private static func step(_ car: RacerCar, dt: Double) -> RacerCar {
        var c = car
        let input = c.lastInput

        // Steering.
        if let i = input {
            c.heading += i.steering * turnRate * dt
        }

        // Throttle / brake.
        if let i = input, i.throttle > 0 {
            c.speed += acceleration * i.throttle * dt
        } else if let i = input, i.brake > 0 {
            c.speed -= braking * i.brake * dt
        } else {
            // Idle decay.
            if c.speed > 0 { c.speed = max(0, c.speed - idleDecay * dt) }
            else if c.speed < 0 { c.speed = min(0, c.speed + idleDecay * dt) }
        }
        c.speed = max(-maxSpeed * 0.5, min(maxSpeed, c.speed))

        // Position update.
        c.x += cos(c.heading) * c.speed * dt
        c.y += sin(c.heading) * c.speed * dt

        // Wall bounce.
        if c.x < 0           { c.x = 0;            c.speed = -c.speed * bounceFactor }
        if c.x > trackWidth  { c.x = trackWidth;   c.speed = -c.speed * bounceFactor }
        if c.y < 0           { c.y = 0;            c.speed = -c.speed * bounceFactor }
        if c.y > trackHeight { c.y = trackHeight;  c.speed = -c.speed * bounceFactor }

        return c
    }
}

// MARK: - Types

public struct RacerInput: Codable, Equatable, Sendable {
    public let player: UUID
    /// 0..1 forward throttle.
    public let throttle: Double
    /// 0..1 brake.
    public let brake: Double
    /// -1..1 steering (negative = left, positive = right).
    public let steering: Double

    public init(player: UUID, throttle: Double, brake: Double, steering: Double) {
        self.player = player
        self.throttle = max(0, min(1, throttle))
        self.brake = max(0, min(1, brake))
        self.steering = max(-1, min(1, steering))
    }
}

public struct RacerCar: Codable, Equatable, Sendable {
    public var x: Double
    public var y: Double
    public var heading: Double // radians
    public var speed: Double
    public var lastInput: RacerInput?

    public init(x: Double, y: Double, heading: Double, speed: Double, lastInput: RacerInput? = nil) {
        self.x = x; self.y = y; self.heading = heading; self.speed = speed; self.lastInput = lastInput
    }
}

public struct RacerState: Codable, Equatable, Sendable {
    public let host: UUID
    public let guest: UUID
    public var cars: [UUID: RacerCar]
    public var laps: [UUID: Int]
    public var tickCount: Int
    /// Set when the race is over. v1 has no automatic finish line; the
    /// lobby will mark a winner manually after a fixed-time race.
    public var winnerId: UUID?
}

File: apps/BuddyPlay/android/core/src/main/java/com/americangroupllc/buddyplay/core/domain/RacerPhysics.kt
package com.americangroupllc.buddyplay.core.domain

import com.americangroupllc.buddyplay.core.models.Peer
import kotlinx.serialization.Serializable
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

/**
 * Pure deterministic 2D top-down racer physics. `tick(state, dt, input)`
 * returns the next state. Identical inputs + identical dt produce identical
 * state on both peers, enabling client-side prediction + host reconciliation.
 *
 * Mirror of `shared/BuddyCore/Sources/BuddyCore/Domain/RacerPhysics.swift`.
 */
object RacerPhysics : GameStateReducer<RacerState, RacerInput> {

    const val TRACK_WIDTH: Double  = 100.0
    const val TRACK_HEIGHT: Double = 60.0
    const val MAX_SPEED: Double    = 30.0
    const val ACCELERATION: Double = 50.0
    const val BRAKING: Double      = 70.0
    const val IDLE_DECAY: Double   = 12.0
    const val TURN_RATE: Double    = 3.5
    const val BOUNCE_FACTOR: Double = 0.55

    override fun initialState(host: Peer, guest: Peer): RacerState =
        RacerState(
            host = host.id,
            guest = guest.id,
            cars = mutableMapOf(
                host.id  to RacerCar(20.0, 30.0, 0.0, 0.0, null),
                guest.id to RacerCar(80.0, 30.0, Math.PI, 0.0, null),
            ),
            laps = mutableMapOf(host.id to 0, guest.id to 0),
            tickCount = 0,
            winnerId = null,
        )

    override fun reduce(state: RacerState, input: RacerInput): GameStateReducer.Step<RacerState> {
        if (state.winnerId != null) throw GameStateReducer.Error.GameOver
        val s = state.copy(cars = HashMap(state.cars))
        s.cars[input.player] = s.cars[input.player]!!.copy(lastInput = input)
        val next = tick(s, dtMillis = 33)
        val outcome = next.winnerId?.let { GameStateReducer.Outcome.Winner(it) }
        return GameStateReducer.Step(next, outcome)
    }

    override fun isFinal(state: RacerState): Boolean = state.winnerId != null

    override fun currentTurn(state: RacerState): String? = null

    fun tick(state: RacerState, dtMillis: Int): RacerState {
        if (state.winnerId != null) return state
        val dt = dtMillis / 1000.0
        val s = state.copy(cars = HashMap(state.cars))
        for ((id, car) in s.cars.toMap()) {
            s.cars[id] = step(car, dt)
        }
        s.tickCount += 1
        return s
    }

    private fun step(car: RacerCar, dt: Double): RacerCar {
        var x = car.x; var y = car.y; var heading = car.heading; var speed = car.speed
        val input = car.lastInput

        if (input != null) heading += input.steering * TURN_RATE * dt

        if (input != null && input.throttle > 0) speed += ACCELERATION * input.throttle * dt
        else if (input != null && input.brake > 0) speed -= BRAKING * input.brake * dt
        else {
            if (speed > 0) speed = max(0.0, speed - IDLE_DECAY * dt)
            else if (speed < 0) speed = min(0.0, speed + IDLE_DECAY * dt)
        }
        speed = max(-MAX_SPEED * 0.5, min(MAX_SPEED, speed))

        x += cos(heading) * speed * dt
        y += sin(heading) * speed * dt

        if (x < 0)            { x = 0.0;          speed = -speed * BOUNCE_FACTOR }
        if (x > TRACK_WIDTH)  { x = TRACK_WIDTH;  speed = -speed * BOUNCE_FACTOR }
        if (y < 0)            { y = 0.0;          speed = -speed * BOUNCE_FACTOR }
        if (y > TRACK_HEIGHT) { y = TRACK_HEIGHT; speed = -speed * BOUNCE_FACTOR }

        return car.copy(x = x, y = y, heading = heading, speed = speed)
    }
}

@Serializable
data class RacerInput(
    val player: String,
    val throttle: Double,
    val brake: Double,
    val steering: Double,
) {
    init {
        require(throttle in 0.0..1.0) { "throttle out of range" }
        require(brake in 0.0..1.0)    { "brake out of range" }
        require(steering in -1.0..1.0) { "steering out of range" }
    }
}

@Serializable
data class RacerCar(
    val x: Double,
    val y: Double,
    val heading: Double,
    val speed: Double,
    val lastInput: RacerInput?,
)

@Serializable
data class RacerState(
    val host: String,
    val guest: String,
    val cars: MutableMap<String, RacerCar>,
    val laps: MutableMap<String, Int>,
    var tickCount: Int,
    var winnerId: String?,
)
